src/signature.rs

```diff
@@ -0,0 +1,198 @@
+    /// are not canonically encoded to a valid `secp256k1` signature.
+    pub unsafe fn from_bytes_unchecked(bytes: [u8; Self::LEN]) -> Self {
+        Self(bytes.into())
+    }
+
+    /// Add a conversion from arbitrary slices into owned
+    ///
+    /// # Safety
+    ///
+    /// This function will not panic if the length of the slice is smaller than
+    /// `Self::LEN`. Instead, it will cause undefined behavior and read random
+    /// disowned bytes.
+    ///
+    /// There is no guarantee the provided bytes will be a valid signature. Internally, some FFI
+    /// calls to `secp256k1` are performed and we might have undefined behavior in case the bytes
+    /// are not canonically encoded to a valid `secp256k1` signature.
+    pub unsafe fn from_slice_unchecked(bytes: &[u8]) -> Self {
+        Self(Bytes64::from_slice_unchecked(bytes))
+    }
+
+    /// Copy-free reference cast
+    ///
+    /// There is no guarantee the provided bytes will fit the field.
+    ///
+    /// # Safety
+    ///
+    /// Inputs smaller than `Self::LEN` will cause undefined behavior.
+    pub unsafe fn as_ref_unchecked(bytes: &[u8]) -> &Self {
+        // The interpreter will frequently make references to keys and values using
+        // logically checked slices.
+        //
+        // This function will avoid unnecessary copy to owned slices for the interpreter
+        // access
+        &*(bytes.as_ptr() as *const Self)
+    }
+}
+
+impl Deref for Signature {
+    type Target = [u8; Signature::LEN];
+
+    fn deref(&self) -> &[u8; Signature::LEN] {
+        self.0.deref()
+    }
+}
+
+impl AsRef<[u8]> for Signature {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl AsMut<[u8]> for Signature {
+    fn as_mut(&mut self) -> &mut [u8] {
+        self.0.as_mut()
+    }
+}
+
+impl From<Signature> for [u8; Signature::LEN] {
+    fn from(salt: Signature) -> [u8; Signature::LEN] {
+        salt.0.into()
+    }
+}
+
+impl fmt::LowerHex for Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::UpperHex for Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::Debug for Signature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::Display for Signature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[cfg(feature = "std")]
+mod use_std {
+    use crate::{Error, Message, PublicKey, SecretKey, Signature};
+
+    use secp256k1::recovery::{RecoverableSignature as SecpRecoverableSignature, RecoveryId};
+    use secp256k1::Secp256k1;
+
+    use std::borrow::Borrow;
+
+    impl Signature {
+        // Internal API - this isn't meant to be made public because some assumptions and pre-checks
+        // are performed prior to this call
+        pub(crate) fn to_secp(&mut self) -> SecpRecoverableSignature {
```

Comment:
[nit] Would it be more idiomatic to use `impl From<Signature> for SecpRecoverableSignature` and other similar library type conversions?

Reply:
I'm avoiding as much as possible exposing the internals (e.g. [secp256k1](https://crates.io/crates/secp256k1)) of the library as part of the public API so we can easily switch the backend in the future.

Since traits are always public, we would need to either re-rexport `SecpRecoverableSignature`, or leave a broken `From` as public. For that reason, it was created as `pub(crate)`

Reply:
Makes sense